### PR TITLE
DemuxFastqs can output Illumina FASTQ Standards.

### PR DESCRIPTION
**Pull request by nh13**
Supporting outputting the Illumina standard read names and output file names for FASTQ output.

### CODE REVIEW

File: src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala [hunk 1]
```diff
@@ -375,10 +388,10 @@ val qualityFormat: Option[FastqQualityFormat] = None,
   private def toWriter(sampleInfo: SampleInfo): DemuxWriter = {
     val sample = sampleInfo.sample
     val isUnmatched = sample.sampleName == UnmatchedSampleId
```
Comment (nh13) [inline src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala]: Not sure where to put this, but I think the unmatched BAM is called `Undetermined_sequence.fastq.gz?

File: src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala [hunk 2]
```diff
@@ -116,8 +116,13 @@ object DemuxFastqs {
     Sample(sampleOrdinal=sampleOrdinal, sampleId=UnmatchedSampleId, sampleName=UnmatchedSampleId, libraryId=UnmatchedSampleId, i7IndexBases=Some(noMatchBarcode))
   }
 
-  private[fastq] def toSampleOutputPrefix(sample: Sample, isUnmatched: Boolean, output: DirPath, unmatched: String): PathPrefix = {
-    if (isUnmatched) PathUtil.removeExtension(output.resolve(unmatched)) else outputPrefixFrom(output, sample)
+  private[fastq] def toSampleOutputPrefix(sample: Sample, isUnmatched: Boolean, illuminaStandards: Boolean, output: DirPath, unmatched: String): PathPrefix = {
+    (isUnmatched, illuminaStandards) match {
+      case (true, true)   => output.resolve(f"${UnmatchedSampleId}_S${sample.sampleOrdinal}_L${sample.lane.getOrElse(1)}%03d")
```
Comment (tfenne) [inline src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala]: Just FYI, that different versions of Illumina software do different sets of character substitutions on the `sample.sampleName` part of this.  Some turn hyphens into underscores, some turn underscores into hyphens.  I think all replace spaces (with either underscores or hyphens).  So just beware that this isn't a perfect match for any illumina software version in all cases.

File: src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala [hunk 3]
```diff
@@ -266,13 +272,14 @@ val qualityFormat: Option[FastqQualityFormat] = None,
...
+ @arg(doc="The platform unit (typically `<flowcell-barcode>-<sample-barcode>.<lane>`)") val platformUnit: Option[String] = None,
  @arg(doc="The sequencing center from which the data originated") val sequencingCenter: Option[String] = None,
  @arg(doc="Predicted median insert size, to insert into the read group header") val predictedInsertSize: Option[Integer] = None,
  @arg(doc="Platform model to insert into the group header (ex. miseq, hiseq2500, hiseqX)") val platformModel: Option[String] = None,
  @arg(doc="Comment(s) to include in the merged output file's header.", minElements = 0) val comments: List[String] = Nil,
  @arg(doc="Date the run was produced, to insert into the read group header") val runDate: Option[Iso8601Date] = None,
- @arg(doc="Output gzipped FASTQs (`.fastq.gz`) instead of BAM files") val outputFastqs: Boolean = false
+ @arg(doc="Output gzipped FASTQs (`.fastq.gz`) instead of BAM files") val outputFastqs: Boolean = false,
+ @arg(doc="Output FASTQs according to Illumina standards, for example, for upload to the BaseSpace Sequence Hub") val illuminaStandards: Boolean = false
```
Comment (tfenne) [inline src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala]: Could `name` or `naming` get added in there somewhere?  Just to make it clear this only effects file names, and not e.g. quality scaling?

File: src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala [hunk 4]
```diff
@@ -282,12 +289,18 @@ val qualityFormat: Option[FastqQualityFormat] = None,
...
   validate(readStructures.flatMap(_.sampleBarcodeSegments).nonEmpty, s"No sample barcodes found in read structures: " + readStructures.map(_.toString).mkString(", "))
 
+
   private val pairedEnd = readStructures.count(_.templateSegments.nonEmpty) match {
     case 1 => false
     case 2 => true
     case n => invalid(s"Found $n read structures with template bases but expected 1 or 2.")
   }
 
+  if (illuminaStandards) {
```
Comment (tfenne) [inline src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala]: Playing devil's advocate, this isn't mentioned in the usage and seems overly mean.  Why can't I get single-end illumina fastqs?

File: src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala [hunk 5]
```diff
@@ -438,16 +451,39 @@ private class SamRecordWriter(output: PathToBam,
...
-      .toIndexedSeq
+private[fastq] class FastqRecordWriter(prefix: PathPrefix, val pairedEnd: Boolean, val illuminaStandards: Boolean = false) extends DemuxWriter {
+  private val writers: IndexedSeq[FastqWriter] = extensions.map { ext =>
+    FastqWriter(Io.toWriter(PathUtil.pathTo(prefix + ext)))
+  }.toIndexedSeq
+
+  private[fastq] def extensions: Seq[String] = (pairedEnd, illuminaStandards) match {
+    case (true, true)   => Seq("_R1_001.fastq.gz", "_R2_001.fastq.gz")
+    case (true, false)  => Seq("_R1.fastq.gz", "_R2.fastq.gz")
+    case (false, true)  => throw new IllegalArgumentException("Paired end reads required with illumina-standards")
```
Comment (tfenne) [inline src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala]: Why?

File: src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala [hunk 6]
```diff
@@ -438,16 +451,39 @@ private class SamRecordWriter(output: PathToBam,
...
+    case (false, false) => Seq(".fastq.gz")
+  }
+
+  private[fastq] def readName(rec: DemuxRecord): String = {
+    if (illuminaStandards) {
+      val comment = rec.comment.getOrElse {
```
Comment (tfenne) [inline src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala]: I think all this validation of the incoming read name should at least be mentioned in the usage.  E.g. changing the option to say something like "Require that the input fastq files contain Illumina format read names with comments containing `ReadNum:FilterFlag:0:SampleNumber`, and create output files with illumina format file names."

File: src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala (hunk 2 above)
Comment (nh13) [reply src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala]: Awesome.

**Pull request merged**